tool/ci/pkg/bot/check.go

```diff
@@ -0,0 +1,225 @@
+package bot
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/tool/ci"
+	"github.com/gravitational/teleport/tool/ci/pkg/environment"
+
+	"github.com/google/go-github/v37/github"
+	"github.com/gravitational/trace"
+)
+
+// Check checks if all the reviewers have approved the pull request in the current context.
+func (c *Bot) Check(ctx context.Context) error {
```

Comment:
I think you need to break up this logic into functions. The top level function that is called should be simple to read without knowing internal implementation details to be able to reason about how it works and security.

Take a look at the example below.

```go
func (c *Bot) Check(ctx context.Context) error {
  // Remove any stale workflow runs. As only the current workflow run should
  // be shown [add explanation why this is the case].
  //
  // Note: This is run for all workflows, but as to be run once again in a
  // cron workflow because [explain why this is the case].
  err := removeStaleWorkflows()
  if err != nil {
     return trace.Wrap(err)
  }
  
  // Check if the assigned reviewers have approved this PR.
  err := checkApprovals()
  if err != nil {
     return trace.Wrap(err)
  }
  
  return nil
}

func (c *Bot) checkApprovals() error {
  // Get reviews attached to the PR as well as the reviewers for the PR and
  // make sure all required reviewers have approved.
  reviews, err := getReviews()
  if err != nil {
     return trace.Wrap(err)
  }
  reviewers, err := getReviewers() {
     return trace.Wrap(err)
  }
  err := isApproved(reviews, reviewers) {
     return trace.Wrap(err)
  }

  // For external contributors, invalidate all approvals if new commits have
  // been pushed since reviewers approved the PR.
  if isExternal() && newCommits() {
     err := dissmissApprovals()
     if err != nil {
        return trace.Wrap(err)
     }
  }
  
  return nil
}
```